internal/controller/managedcluster_controller.go

```diff
@@ -683,6 +695,357 @@ func (r *ManagedClusterReconciler) machinesAvailable(ctx context.Context, namesp
+	if err != nil {
+		managedCluster.Status.CCMSecretReconciled = false
+		return ctrl.Result{},
+			fmt.Errorf("failed to get cluster providers for cluster %s/%s: %s", managedCluster.Namespace, managedCluster.Name, err)
+	}
+
+	kubeconfSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Name:      fmt.Sprintf("%s-kubeconfig", managedCluster.Name),
+		Namespace: managedCluster.Namespace,
+	}, kubeconfSecret); err != nil {
+		managedCluster.Status.CCMSecretReconciled = false
+		return ctrl.Result{},
+			fmt.Errorf("failed to get kubeconfig secret for cluster %s/%s: %s", managedCluster.Namespace, managedCluster.Name, err)
+	}
+
+	for _, provider := range providers {
+		switch provider {
+		case "aws":
+			l.Info("Skipping creds propagation for AWS")
+			continue
+		case "azure":
+			l.Info("Azure creds propagation start")
+			err := r.propagateAzureSecrets(ctx, managedCluster, kubeconfSecret)
+			if err != nil {
+				managedCluster.Status.CCMSecretReconciled = false
+				errMsg := fmt.Sprintf("failed to create Azure CCM credentials: %s", err)
+				apimeta.SetStatusCondition(managedCluster.GetConditions(), metav1.Condition{
+					Type:    hmc.CredentialsPropagatedCondition,
+					Status:  metav1.ConditionFalse,
+					Reason:  hmc.FailedReason,
+					Message: errMsg,
+				})
+				return ctrl.Result{}, errors.New(errMsg)
+			}
+			apimeta.SetStatusCondition(managedCluster.GetConditions(), metav1.Condition{
+				Type:    hmc.CredentialsPropagatedCondition,
+				Status:  metav1.ConditionTrue,
+				Reason:  hmc.SucceededReason,
+				Message: "Azure CCM credentials created",
+			})
+			continue
+		case "vsphere":
+			l.Info("vSphere creds propagation start")
+			err := r.propagateVSphereSecrets(ctx, managedCluster, kubeconfSecret)
+			if err != nil {
+				managedCluster.Status.CCMSecretReconciled = false
+				errMsg := fmt.Sprintf("failed to create vSphere CCM credentials: %s", err)
+				apimeta.SetStatusCondition(managedCluster.GetConditions(), metav1.Condition{
+					Type:    hmc.CredentialsPropagatedCondition,
+					Status:  metav1.ConditionFalse,
+					Reason:  hmc.FailedReason,
+					Message: errMsg,
+				})
+				return ctrl.Result{}, errors.New(errMsg)
+			}
+			apimeta.SetStatusCondition(managedCluster.GetConditions(), metav1.Condition{
+				Type:    hmc.CredentialsPropagatedCondition,
+				Status:  metav1.ConditionTrue,
+				Reason:  hmc.SucceededReason,
+				Message: "vSphere CCM credentials created",
+			})
+			continue
+		default:
+			managedCluster.Status.CCMSecretReconciled = true
+			errMsg := fmt.Sprintf("unsupported infrastructure provider %s", provider)
+			apimeta.SetStatusCondition(managedCluster.GetConditions(), metav1.Condition{
+				Type:    hmc.CredentialsPropagatedCondition,
+				Status:  metav1.ConditionFalse,
+				Reason:  hmc.FailedReason,
+				Message: errMsg,
+			})
+			return ctrl.Result{}, errors.New(errMsg)
```

Comment:
i think such functions/methods returns should be given without errors because it'll just clog the event queue along with logs with unnecessary same events/messages, I think setting the conditions is good enough 

applicable elsewhere if any

Reply:
@zerospiel That's what I meant when put this note above 
> I'm not sure if we should fail on unsupported infrastructure provider.

I can just skip return _here_ completely. 

If you mean that `reconcileCredentialPropagation` method shouldn't re-trigger reconcile then I don't understand it and it makes not sense to me.

Reply:
Correct, skip the error return only here, nothing else has been meant since there is no more such code that'd behave in the same way